Merge upstream, add some new UI elements 09/13

JashoBell opened the pull request:
This pull request includes a merge of all upstream changes as of ~8pm UTC 09/13.

I made some modest modifications to the plugin to incorporate the negative prompt feature, and threw in default arguments to the txt2img/img2img calls so that they still run. So far as I can tell, everything works. As I have time going forward I will try to incorporate the other new features.

You may already have done these things, but if not I figured it would save some time. Please let me know if I can make any adjustments or have borked anything- this is my first PR 😅 

Thank you for the plugin by the way, it's been an amazing tool.

sddebz commented on krita_server.py:
```diff
@@ -226,8 +233,12 @@ async def f_img2img(req: Img2ImgRequest):
+        req.prompt or collect_prompt(opt, 'prompts'),
+        req.negative_prompt or collect_prompt(opt, 'negative_prompt'),
+        "None",
         image,
+        mask,
+        1,
         {"image": image, "mask": mask},
```
I think the order of arguments is a bit wrong here. Probably should be like this. It matters for inpainting mode.
```suggestion
        image,
        {"image": image, "mask": mask},
        mask,
        1,
```

sddebz commented on krita_server.py:
```diff
@@ -240,18 +251,23 @@ async def f_img2img(req: Img2ImgRequest):
         req.upscale_overlap or opt['upscale_overlap'],
         req.inpaint_full_res or opt['inpaint_full_res'],
-        0,
+        req.mask_invert or opt['mask_invert'],
```
I'm not sure that will work. The problem is that plugin saves the image with the mask applied already. The server doesn't even get masked part of the image. I think some changes in krita_diff.py are needed for this thing to work. 

However I'm fine if this option ends in the master. It can be fixed later. Probably better to hide it though.

sddebz commented on krita_plugin/krita_diff/krita_diff.py:
```diff
@@ -149,6 +151,7 @@ def img2img(self, path, mask_path, mode):
             "max_size": self.cfg('img2img_max_size', int),
             "seed": self.cfg('img2img_seed', str) if not self.cfg('img2img_seed', str).isspace() else '',
             "tiling": tiling,
+            "invert_mask": self.cfg('img2img_invert_mask', bool),
```
This option is called `mask_invert` in krita_server.py. So it doesn't really work.

sddebz commented on krita_server.py:
```diff
@@ -115,6 +116,7 @@ class Img2ImgRequest(BaseModel):
     inpainting_fill: Optional[int]
     inpaint_full_res: Optional[bool]
     mask_blur: Optional[int]
+    mask_invert: Optional[bool]
```
This option is called `invert_mask` in krita_diff.py

sddebz commented on requirements_versions.txt:
```diff
@@ -1,13 +1,18 @@
 omegaconf==2.1.1
 pytorch_lightning==1.7.2
-pyyaml==6.0
-uvicorn==0.18.3
-fastapi==0.81.0
```
These two are requirements for krita_server.py. They should not be removed.

sddebz commented on krita_plugin/krita_diff/krita_diff_ui.py:
```diff
@@ -235,9 +236,14 @@ def create_img2img_interface(self):
         self.img2img_prompt_text = QPlainTextEdit()
         self.img2img_prompt_text.setPlaceholderText("krita_config.yaml value will be used")
-        self.img2img_prompt_layout = QHBoxLayout()
+        self.img2img_negative_prompt_label = QLabel("Negative Prompt:")
```
I think it's worth adding this option for txt2img too. But it's fine anyway.

sddebz commented on README.md:
```diff
@@ -6,6 +6,43 @@ Requires Krita 5.1
 ## Usage example
 [target.webm](https://user-images.githubusercontent.com/112324253/188291339-9d146a9a-ba9f-4671-9bd8-c8b55fd48ba6.webm)
 
+[Detailed feature showcase with images, art by Greg Rutkowski](https://github.com/AUTOMATIC1111/stable-diffusion-webui-feature-showcase)
```
Generally there are many things that just not in the Krita plugin. It's a bit misleading. Like there is no X/Y plot, no colab support, no variations. I would prefer not to copy README from upstream, and just link upstream instead. 

The only thing worth change would be installation instructions. I'll update it anyway soon.

sddebz commented on krita_plugin/krita_diff/krita_diff.py:
```diff
@@ -137,8 +139,8 @@ def img2img(self, path, mask_path, mode):
             "mode": mode,
             "src_path": path,
             "mask_path": mask_path,
-            "prompt": self.fix_prompt(
-                self.cfg('img2img_prompt', str) if not self.cfg('img2img_prompt', str).isspace() else None),
+            "prompt": self.cfg('img2img_prompt', str) if not self.cfg('img2img_prompt', str).isspace() else None,
```
I think it's worth to leave fix_prompt function here. This function replaces new lines with commas and removes redundant empty lines from prompt. I don't think there is a real difference for a model, but it makes logs more readable.

sddebz reviewed (commented):
Thank you very much. The repository is lagging behind upstream, so this PR is very welcome. 

I have some comments. The serious one is about arguments for img2img. If this one is fixed, I'll merge this PR.

JashoBell commented:
Thank you for the quick feedback and noting my errors, I always feel as if I'm being thorough but manage to overlook something. I think I've addressed your above comments in the last commit.

sddebz commented:
Ok, it works on my machine. I merged it. Thank you.
